### DISCUSSION TITLE
Release v0.2.0

### DISCUSSION THREAD

**Discussion by rine77** [rine77/homeassistantedupage#29] [Announcements]
- new config flow to select student
- multiple students in multiple accounts implemented
- timetables now available as calendars with lessons as events
- lessons are loaded 14 days ahead
- teachers and rooms inside events in calendar
- update timer set to 30min but will be configurable in future releases

<hr /><em>This discussion was created from the release <a href='https://github.com/rine77/homeassistantedupage/releases/tag/v0.2.0'>Release v0.2.0</a>.</em>

**Comment by rine77** [comment]
I installed it in my production via HACS after removing everything and it works like a charme.
Any tests appreciated!

**Reply by brennyB** [reply]
Hi Rine77, 

works out of the box in my installation.

updated via HACS,
restartet HomeAssistant
reconfigured (logins and Subdomain...)

done! 

THX

**Comment by mgrfilipmarek** [comment]
Multiple issues at my integration:

1. **not unique ID**

Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:764
integration: Sensor (documentation, issues)
First occurred: 6:49:54 AM (1 occurrences)
Last logged: 6:49:54 AM
Platform homeassistantedupage does not generate unique IDs. ID edupage_grades_1070191_náboženská_výchova already exists - ignoring sensor.juraj_marek_nabozenska_vychova


2. **List index out of range**

This error originated from a custom integration.
Logger: custom_components.homeassistantedupage.homeassistant_edupage
Source: custom_components/homeassistantedupage/homeassistant_edupage.py:40
integration: homeassistantedupage (documentation, issues)
First occurred: 6:49:51 AM (2 occurrences)
Last logged: 6:49:54 AM
EDUPAGE unexpected login error: list index out of range


3. **Error on device update**

Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:728
integration: Sensor (documentation, issues)
First occurred: 6:49:58 AM (1 occurrences)
Last logged: 6:49:58 AM

homeassistantedupage: Error on device update!
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 728, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1302, in async_device_update
    await self.async_update()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 599, in async_update
    await self.coordinator.async_request_refresh()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 273, in async_request_refresh
    await self._debounced_refresh.async_call()
  File "/usr/src/homeassistant/homeassistant/helpers/debounce.py", line 114, in async_call
    await task
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 358, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 481, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 176, in async_update_listeners
    update_callback()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 561, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1007, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    self.__async_calculate_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1073, in __async_calculate_state
    if extra_state_attributes := self.extra_state_attributes:
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/homeassistantedupage/sensor.py", line 74, in extra_state_attributes
    attributes[f"grade_{i+1}_teacher"] = grade.teacher.name
                                         ^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'name'


4. **Error adding entity**

Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:918
integration: Sensor (documentation, issues)
First occurred: 6:49:58 AM (1 occurrences)
Last logged: 6:49:58 AM

Error adding entity sensor.juraj_marek_vytvarna_vychova for domain sensor with platform homeassistantedupage
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 918, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1368, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1007, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    self.__async_calculate_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1073, in __async_calculate_state
    if extra_state_attributes := self.extra_state_attributes:
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/homeassistantedupage/sensor.py", line 74, in extra_state_attributes
    attributes[f"grade_{i+1}_teacher"] = grade.teacher.name
                                         ^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'name'


5. **warning - awaited result**

Logger: py.warnings
Source: config_entries.py:635
First occurred: 6:49:58 AM (1 occurrences)
Last logged: 6:49:58 AM

/usr/src/homeassistant/homeassistant/config_entries.py:635: RuntimeWarning: coroutine 'Edupage.login' was never awaited result = await component.async_setup_entry(hass, self)

**Reply by rine77** [reply]
please try v0.2.1

**Comment by rine77** [comment]
thank you very much for the feedback!! Your logfile is pretty helpful, thanks!

First things first: try to completely remove the integration, including any file on HA filesystem:

1. remove with UI
2. remove all sensors belonging to the homeassistantedupage integration
3. use file system terminal to go to the folder /homeassistant/custom_components/ and delete all the following folders: homeassistant-edupage, homeassistantedupage, homeassistant_edupage
4. remove it from HACS
5. "downloaded" -> 2 dots on the right side -> remove
6. restart HA
7. add it to HACS as mentioned in Readme

After the re-installation I am interested in the new logfile.

Problems in detail:

- Complete removal should help at least with the unique_id problem (1)
- Problem (2) can safely be ignored at the moment. It doesnt change functionality and is known to me.
- Number 3 and number 4 are really nice bugs to me because they are related to your content. I am on it and will create an ISSUE for that. It will be fixed in v0.2.1 very soon.
- Number 5 is okay at the moment, its a warning only and to ignore it helped me to release 0.2 yesterday ;)

Best regards
/rine

**Comment by mgrfilipmarek** [comment]
Im on v0.2.1 currently (removed old one completely and reinstalled to new one):

Calendar with subjects not working correctly:
![image](https://github.com/user-attachments/assets/6d040f49-d60a-4a0b-b718-c5a7de95293b)

3 things in log:
![image](https://github.com/user-attachments/assets/752dff00-eac6-4e7b-97bf-eea4b922dd28)

**Reply by rine77** [reply]
are you sure there are Timetable Entries in your Edupage instance?
Likewise:
![image](https://github.com/user-attachments/assets/d615a0f9-d215-4fe1-9490-5ab560428fc3)

**Reply by mgrfilipmarek** [reply]
yes, its there.....but looks like its there only as an image 😢 
![image](https://github.com/user-attachments/assets/37fc8b29-c99c-44ab-a20e-2b4fe070ec2a)

**Reply by rine77** [reply]
No I think its data, not an image. In UI it behaves like an image because of the mouse pointer. Aweful UI anyways...

BUT you are absolutely sure that you have removed not only the integration before upgrading but also EVERY sensor and calendar regarding my integration? Yeah it sucks because there are a lot...

**Reply by mgrfilipmarek** [reply]
How to double check if there is any sensor/calendar left after uninstall?

**Reply by rine77** [reply]
after you removed the integration completely, check under "settings" -> "devices&services" -> "Entities" which entites are abandoned (marked with a red circle in the "state" column).
Remove all of them before reinstalling. Yeah maybe there are a lot, but in next release I will implement a function to clean up everything automatically.
After really everything is gone - please reinstall and check the logs/functionality.

**Reply by rine77** [reply]
after I did it that way I have this error left (twice) but it can be tolerated.
Calendars and Subject-Entities are being loaded smoothly and theres no unique_id error or something:

![image](https://github.com/user-attachments/assets/dc1118a7-835e-4146-bf1c-06ad4ab9a39a)

**Reply by mgrfilipmarek** [reply]
After reinstall, Im in the same state with integration - 3things in log and not working calendar (so no news here).
Maybe I can try to setup integration / dev environment locally in some IDE (I`m java dev 😃).

**Reply by rine77** [reply]
Awesome! So maybe you are familiar with vscode? I use it as IDE and you can find a .vscode folder in root directory. There is a launch file. Furthermore you need

- homeassistant core git repository (thats the local instance I use to develop)
- python at least 3.12 (of course)
- venv (easy to create in vscode)
- 
I cant get into detail very much cause it would create a wall of text but maybe it helps. As a dev I guess youre fine with this.

In the other way around I wanted to ask you to give me your credentials temporarily and I could connect to your Edupage instance. If you can imagine doing that, I sent you my pgp public key. If not - no offense. I wouldnt do it :)

/rine

**Reply by mgrfilipmarek** [reply]
@rine77 Can you ping me please?

**Comment by mgrfilipmarek** [comment]
In v0.2.2, sensors names are damaged - looks like encoding issue.
Correct is (in slovak lang. "voľnočasové aktivity")
![image](https://github.com/user-attachments/assets/0bc86fff-7b8f-4b5f-89fd-3a0acd017e08)

**Reply by rine77** [reply]
yeah my fault, try bugfix https://github.com/rine77/homeassistantedupage/releases/tag/v0.2.3

Please tell me if you still have lots of zombie sensors or if they disapper and change their names (because I dont touched the unique_id it should work)

**Reply by mgrfilipmarek** [reply]
ok, name looks better but entity id not:
![image](https://github.com/user-attachments/assets/bacb0ce9-40d9-4e05-83a5-301a96fda630)

In older screen we can see that interpunction chars were replaced correctly:
https://github.com/rine77/homeassistantedupage/issues/21#issuecomment-2493863710

**Comment by mgrfilipmarek** [comment]
In v0.2.4, encoding issue is solved 😃

**Comment by mgrfilipmarek** [comment]
In v0.2.5 - I cant login anymore. I have edupage account without 2FA (@bbirnbach maybe you can check, please?)

Logger: homeassistant.config_entries
Source: config_entries.py:635
First occurred: 10:48:19 AM (2 occurrences)
Last logged: 10:48:42 AM

Error setting up entry Edupage (Juraj Marek) for homeassistantedupage
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 635, in __async_setup_with_context
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/homeassistantedupage/__init__.py", line 29, in async_setup_entry
    PHPSESSID = entry.data[CONF_PHPSESSID]
                ~~~~~~~~~~^^^^^^^^^^^^^^^^
KeyError: 'PHPSESSID'


This error originated from a custom integration.

Logger: custom_components.homeassistantedupage.config_flow
Source: custom_components/homeassistantedupage/config_flow.py:85
integration: homeassistantedupage (documentation, issues)
First occurred: 10:49:44 AM (2 occurrences)
Last logged: 10:50:10 AM

Exception during API call: 'NoneType' object has no attribute 'is_confirmed'

**Reply by rine77** [reply]
I renamed 0.2.5. 
Please use 0.2.4 if you dont have 2FA.

**Reply by bbirnbach** [reply]
> In v0.2.5 - I cant login anymore.

Have you tried to delete and readd the integration?
Because of the 2FA, changes were needed in the login process --> you have to login again.

**Reply by rine77** [reply]
0.2.6 should solve all login problems. Please test and report :)

**Reply by mgrfilipmarek** [reply]
all seems to be correct 😃
